File: app/src/main/java/xyz/ummo/user/adapters/ServicesAdapter.java
package xyz.ummo.user.adapters;

import android.content.Context;
import android.content.Intent;
import android.preference.PreferenceManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import timber.log.Timber;
import xyz.ummo.user.AgentRequest;
import xyz.ummo.user.ui.detailedService.DetailedProduct;
import xyz.ummo.user.R;
import xyz.ummo.user.Service;
import xyz.ummo.user.delegate.DelegateService;
import xyz.ummo.user.delegate.User;

public class ServicesAdapter extends RecyclerView.Adapter<ServicesAdapter.MyViewHolder>  {

    private List<Service> services;
    Context context;
    private String departmentName;
    private static final String TAG = "ServicesAdapter";

    public class MyViewHolder extends RecyclerView.ViewHolder {

        TextView serviceTitle, serviceDescription, serviceForm, servicePersonalDocs,
                serviceCost, serviceDuration, moreButton;
        String steps = "";

        Button requestAgentButton;

        public RelativeLayout departmentGround;

        MyViewHolder(View view) {
            super(view);
            //departmentGround = view.findViewById(R.id.department_ground);
            serviceTitle= view.findViewById(R.id.service_title);
            serviceDescription = view.findViewById(R.id.service_description);
            serviceForm = view.findViewById(R.id.service_form);
            servicePersonalDocs = view.findViewById(R.id.service_personal_docs);
            serviceCost = view.findViewById(R.id.service_cost);
            serviceDuration = view.findViewById(R.id.service_duration);

            requestAgentButton = view.findViewById(R.id.request_agent_btn);

            moreButton = view.findViewById(R.id.more_on_the_service_text);
        }
    }


    public ServicesAdapter(Context context, List<Service>  services, String departmentName) {
        this.services = services;
        this.context = context;
        this.departmentName = departmentName;
    }

    @NonNull
    @Override
    public ServicesAdapter.MyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.service_content, parent, false);

        Log.e("Adapter","create vh");
        return new ServicesAdapter.MyViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ServicesAdapter.MyViewHolder holder, int position) {
        Service service = services.get(position);

        StringBuilder steps = new StringBuilder();

        for (int i = 0; i < service.getSteps().size(); i++){
            steps.append(service.getSteps().get(i)).append(" ");
        }
        holder.steps = steps.toString();

        holder.serviceTitle.setText(service.getServiceName());
        holder.serviceDescription.setText(service.getServiceDescription());
        holder.serviceForm.setText(" " + steps);
        holder.servicePersonalDocs.setText(" " + service.getPersonalDocs());
        holder.serviceCost.setText(" " + service.getCost());
        holder.serviceDuration.setText( " " + service.getDuration());


        final String serviceName, description, form, personalDocs, cost, duration;

        serviceName = holder.serviceTitle.getText().toString();
        description = holder.serviceDescription.getText().toString();
        form = holder.serviceForm.getText().toString();
        personalDocs = holder.servicePersonalDocs.getText().toString();
        cost = holder.serviceCost.getText().toString();
        duration = holder.serviceDuration.getText().toString();

        holder.requestAgentButton.setOnClickListener(view -> {

            Intent i= new Intent(context, AgentRequest.class);
            i.putExtra("departmentName", departmentName);
          //  i.putExtra("serviceId",service)
            i.putExtra("serviceName", serviceName);
            i.putExtra("form", form);
            i.putExtra("id",service.getId());
            i.putExtra("docs", personalDocs);
            i.putExtra("cost", cost);
            i.putExtra("steps",holder.steps);
            i.putExtra("duration", duration);
             String jwt = PreferenceManager.getDefaultSharedPreferences(ServicesAdapter.this.context).getString("jwt", "");
            new DelegateService(ServicesAdapter.this.context, User.Companion.getUserId(jwt),service.getId()){
                @Override
                public void done(@NotNull byte[] data, int code) {
                    Timber.e(new String(data));
                }
            };
           // context.startActivity(i);
        });

        holder.moreButton.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View v) {

                Intent intent = new Intent(context, DetailedProduct.class);
                intent.putExtra("serviceName", serviceName);
                intent.putExtra("description", description);
                intent.putExtra("cost", cost);
                intent.putExtra("steps",holder.steps);
                intent.putExtra("duration", duration);
                intent.putExtra("docs",personalDocs);
                Timber.e("onClick: MORE-BUTTON clicked!");
                context.startActivity(intent);
            }
        });
    }

    @Override
    public int getItemCount() {
        return services.size();
    }
}

File: app/src/main/java/xyz/ummo/user/adapters/ServicesAdapter.kt
package xyz.ummo.user.adapters

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.mixpanel.android.mpmetrics.MixpanelAPI
import kotlinx.android.synthetic.main.service_slice.view.*
import org.json.JSONObject
import xyz.ummo.user.R
import xyz.ummo.user.data.entity.ServiceEntity
import xyz.ummo.user.ui.detailedService.DetailedServiceActivity
import xyz.ummo.user.ui.fragments.bottomSheets.ServiceOptionsMenuBottomSheet
import xyz.ummo.user.utilities.PARENT
import xyz.ummo.user.utilities.SERVICE_ENTITY
import java.io.Serializable


class ServicesAdapter(private var optionsMenuClickListener: OptionsMenuClickListener) :
    RecyclerView.Adapter<ServicesAdapter.ServiceViewHolder>() {

    private lateinit var mContext: Context

    private lateinit var mixpanel: MixpanelAPI

    interface OptionsMenuClickListener {
        fun onOptionsMenuClicked(position: Int)
    }

    inner class ServiceViewHolder(serviceView: View) : RecyclerView.ViewHolder(serviceView)

    private val differCallback = object : DiffUtil.ItemCallback<ServiceEntity>() {
        override fun areItemsTheSame(oldItem: ServiceEntity, newItem: ServiceEntity): Boolean {
            return oldItem.serviceId == newItem.serviceId
        }

        override fun areContentsTheSame(oldItem: ServiceEntity, newItem: ServiceEntity): Boolean {
            return oldItem == newItem
        }
    }

    val differ = AsyncListDiffer(this, differCallback)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ServiceViewHolder {
        mContext = parent.context
        mixpanel = MixpanelAPI.getInstance(
            mContext,
            mContext.resources.getString(R.string.mixpanelToken)
        )

        return ServiceViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.service_slice,
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(holder: ServiceViewHolder, position: Int) {
        val serviceEntity = differ.currentList[position]

        holder.itemView.apply {
            service_title_text_view_slice.text = serviceEntity.serviceName
            service_description_text_view_slice.text = serviceEntity.serviceDescription
            setOnClickListener {
                onItemClickListener?.let {
                    it(serviceEntity)
                }
            }

            if (serviceEntity.delegatable!!) {
                service_delegatable_tag_relative_layout.visibility = View.VISIBLE
            }

            options_menu_service_slice.setOnClickListener {
//                optionsMenuClickListener.onOptionsMenuClicked(position)
                val serviceBundle = Bundle()
                val serviceEntityObject = JSONObject()
                val serviceOptionsMenuBottomSheet = ServiceOptionsMenuBottomSheet()
                serviceBundle.putSerializable(SERVICE_ENTITY, serviceEntity)
                serviceOptionsMenuBottomSheet.arguments = serviceBundle
                serviceOptionsMenuBottomSheet.show(
                    (mContext as FragmentActivity).supportFragmentManager,
                    ServiceOptionsMenuBottomSheet.TAG
                )

                serviceEntityObject.put("service_name", serviceEntity.serviceName)
                mixpanel.track("Service Options Menu", serviceEntityObject)
            }

            /** Let's let the user tap on any item on the view to launch more info **/
            /*open_service_image_slice.setOnClickListener {
                showServiceDetails(serviceEntity)
            }*/
            mini_service_card_view.setOnClickListener {
                showServiceDetails(serviceEntity)
            }
            service_card_avatar_image_view_slice.setOnClickListener {
                showServiceDetails(serviceEntity)
            }
            service_title_text_view_slice.setOnClickListener {
                showServiceDetails(serviceEntity)
            }
            service_description_text_view_slice.setOnClickListener {
                showServiceDetails(serviceEntity)
            }
        }
    }

    /** When a service is tapped on, it should expand to a detailed view with more info. **/
    private fun showServiceDetails(serviceEntity: ServiceEntity) {
        val intent = Intent(mContext, DetailedServiceActivity::class.java)
        intent.putExtra(PARENT, "ServicesAdapter")
        val serviceResults = JSONObject()
        serviceResults.put("Service", intent.getStringExtra(SERVICE_ENTITY))
        mixpanel.track("Found Results", serviceResults)

        /** Passing Service to [DetailedServiceActivity] via [Serializable] object **/
        intent.putExtra(SERVICE_ENTITY, serviceEntity)

        mContext.startActivity(intent)
    }

    private var onItemClickListener: ((ServiceEntity) -> Unit)? = null

    fun setOnItemClickListener(listener: (ServiceEntity) -> Unit) {
        onItemClickListener = listener
    }

    override fun getItemCount(): Int {
        return differ.currentList.size
    }
}